staging/src/k8s.io/apiserver/pkg/storage/etcd3/store.go

```diff
@@ -452,6 +452,19 @@ func getNewItemFunc(listObj runtime.Object, v reflect.Value) func() runtime.Obje
 
 func (s *store) Count(key string) (int64, error) {
 	key = path.Join(s.pathPrefix, key)
+
+	// Consider the following keys:
+	// A: <storage-prefix>//foo.bar.io/machines
+	// B: <storage-prefix>//foo.bar.io/machinesets
+	// If we ask for the count of key A, the result will also include the count for key B
+	// since key B shares the same prefix as key A.
+	//
+	// Add a separator to mark the end of the key here, this will exclude all other
+	// keys from a different resource that is a prefix of the specified key.
+	if len(key) > 0 && key[len(key) - 1] != '/' {
```

Comment:
Great finding - those this deserves a test to ensure we won't regress in the future.

Reply:
Thanks @wojtek-t. I am working on a test. I wanted to make sure this is a good fix.

Reply:
@wojtek-t, just added a test, can you please take a look?